### DISCUSSION TITLE
[1.1.445] IPAdapter Advanced Weighting [Instant Style]

### DISCUSSION THREAD

**Discussion by huchenlei** [Mikubill/sd-webui-controlnet#2770] [General]
IPAdapter advanced weighting control was first implemented in https://github.com/cubiq/ComfyUI_IPAdapter_plus. Now we also have it supported in sd-webui-controlnet!

## News
- [2024-04-29] Academia SD published an awesome tutorial video on ipadapter advanced weighting: https://www.youtube.com/watch?v=r8Kn0KCSKMo

## SDXL composition [instant style]
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/bb4b97ab-51e3-434d-bae4-f7c0901a7fb5)
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/7b7b64fb-d485-4543-93c1-fedb145f65c5)
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/84051669-7fef-4ae7-a423-be459f825ad0)

Infotext:
```
1girl, absurdres, highres, (masterpiece:1.2), (best quality, highest quality),
Negative prompt: (lowres, low quality, worst quality:1.2), (text:1.2), watermark, (frame:1.2), deformed, ugly, deformed eyes, blur, out of focus, blurry, deformed cat, deformed, photo, anthropomorphic cat, monochrome, photo, pet collar, gun, weapon, blue, 3d, drones, drone, buildings in background, green
Steps: 40, Sampler: DPM++ 2M, Schedule type: Karras, CFG scale: 7, Seed: 1058757959, Size: 1000x1000, Model hash: aeb7e9e689, Model: juggernautXL_v8Rundiffusion, Clip skip: 2, ControlNet 0: "Module: ip-adapter-auto, Model: ip-adapter_sdxl [d5d53548], Weight: 1, Resize Mode: Crop and Resize, Low Vram: False, Processor Res: 512, Threshold A: 0.5, Threshold B: 0.5, Guidance Start: 0, Guidance End: 1, Pixel Perfect: False, Control Mode: Balanced, Hr Option: Both, Save Detected Map: True", Version: v1.9.0-15-gff6f4680
```

## SDXL style transfer [instant style]
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/a5f515ad-07d0-4d97-af9e-109baadbce6f)
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/125a6142-d35c-4cac-9e7a-b05dbea62ebd)

Infotext:
```
1girl, absurdres, highres, (masterpiece:1.2), (best quality, highest quality),
Negative prompt: (lowres, low quality, worst quality:1.2), (text:1.2), watermark, (frame:1.2), deformed, ugly, deformed eyes, blur, out of focus, blurry, deformed cat, deformed, photo, anthropomorphic cat, monochrome, photo, pet collar, gun, weapon, blue, 3d, drones, drone, buildings in background, green
Steps: 40, Sampler: DPM++ 2M, Schedule type: Karras, CFG scale: 7, Seed: 1058757959, Size: 1000x1000, Model hash: aeb7e9e689, Model: juggernautXL_v8Rundiffusion, Clip skip: 2, ControlNet 0: "Module: ip-adapter-auto, Model: ip-adapter_sdxl [d5d53548], Weight: 1, Resize Mode: Crop and Resize, Low Vram: False, Processor Res: 512, Threshold A: 0.5, Threshold B: 0.5, Guidance Start: 0, Guidance End: 1, Pixel Perfect: False, Control Mode: Balanced, Hr Option: Both, Save Detected Map: True", Version: v1.9.0-15-gff6f4680
```

## Free edit on weights
You can also freely adjust any weight number by directly editing the weight array.
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/ff00919b-8bdf-4890-b6f0-556396680108)
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/4514ef96-e4e5-43df-9699-35208fb630c5)

## API usage
Just set `advanced_weighting` field in the API to control weight on each block with cross attention. You can refer to `scripts/ipadapter/weight.py` for weight presets used on the UI.

```python
{
  "image": realistic_girl_face_img,
  "module": "ip-adapter-auto",
  "model": "ip-adapter_sd15",
  "advanced_weighting": [1, 1, 1, ..., 1],
}
```

## Note
You are welcome to post your generations with different ipadapter weight settings in this discussion thread.

**Comment by CCpt5** [comment]
This is incredible! I've used the Comfy implementation which is great, but it seems to really help/work well in A1111. Thanks so much for your continued efforts!!

**Comment by IPv6** [comment]
Really cool, thanks!
(sorry to ask - can it be used in Forge too?)

**Reply by huchenlei** [reply]
Unfortunately lllyasviel did not give us collaborators merge permission in forge repo, so any work need to wait until he comes back.

**Reply by Gushousekai195** [reply]
Why don't I see that happening anytime soon???

**Reply by yamfun** [reply]
is it in Forge nowadays?

**Comment by INTstinkt** [comment]
Hey there,
just wanted to ask if there is any kind of documentation about each different weight in the transformer index. For now i mostly found that Output block 6 is mostly for style and Input Block 3 mostly for Composition. I tried to understand what each Block does by 0 ing the block but other then those 2 mentioned Blocks the generation is just a tad different each time. 

Cheers

**Reply by ghost** [reply]
It's more a style management than a composition management. The constant is the blocks are a progression from in00 to out11 in the image construction. By applying math formula to the blocks weights you can manage precisely the degrees of similarity in one way or other. It's like the a1111 use to merge two checkpoint at 25%, 50%, 75% to chose the power of the merge, When you have access to these individual weights you can apply any formula you want not just a simple 25% or 50%, You can create a cosine, a reverse cosine or any formula you want.  .. You can also use DAAM to help a bit to know what the layers the ia used to create a particular image and chose to use just these layers for the final image by example but these layers will work differently for a portrait, a full body or a landscape. The thing you can really use is to apply various formulas to the 25 layers to merge the styles. Also for more info lora has 25 layers. Lycoris has more layers than lora, ect.. Another important point these layer are just the surface of the full prompt, It seem also we have 25 layers per word in the prompt but currently I never see an extension that also give access to all these sub layers. I think having access to these sub layers per word give a more composition approach than just applying the formula to the full prompt that is more a style approach. With DAAM you can use a different prompt in each layer, it's interesting but do not currently give a full access to the sub layers. To resume, the layers for the full prompt is the style and the sub layers per word is the composition. We have access just to the style currently.

**Comment by ghost** [comment]
Normally it's 25 blocks (12 in, 12 outs and one mid). Here you use less 25 ?

**Reply by huchenlei** [reply]
Here indexes are transformer index, not block index. Only a subset of blocks have cross attn (Transformer) in the UNet.

**Reply by huchenlei** [reply]
See https://github.com/Mikubill/sd-webui-controlnet/blob/59d5998823daabd73d6339b3d21c9cdda27b0286/scripts/enums.py#L87-L146 for ref

**Reply by huchenlei** [reply]
BTW, 25 blocks is only for SD15/SD21, SDXL only has 19 blocks (9 in, 9 out and 1 mid)

**Comment by hrz009** [comment]
I'm using Stable Diffusion 1.5 and after this update, ip adapter is not working. Can someone explain how to use it for SD1.5 and not SDXL. All the pics generated from ipadapter are just green blurry images

**Reply by TheSlayonetta** [reply]
i think it's specifically for SDXL... I'm going to revert back to an old repository until it gets sorted cause it has become unusable for me with this feature, so you're not alone

**Reply by huchenlei** [reply]
The advanced weighting is for both SD15 and SDXL. Please file an issue to reproduce your issue, thanks!

**Reply by hrz009** [reply]
How to file an issue, can you please explain.

**Reply by huchenlei** [reply]
https://github.com/Mikubill/sd-webui-controlnet/issues/new/choose

**Comment by AhBumm** [comment]
This feature seems can't work with Multi-Inputs. report:
 `ValueError: controlnet is enabled but no input image is given`

**Reply by AhBumm** [reply]
```
*** Error running process: E:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\controlnet.py
    Traceback (most recent call last):
      File "E:\stable-diffusion-webui\modules\scripts.py", line 825, in process
        script.process(p, *script_args)
      File "E:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\controlnet.py", line 1247, in process
        self.controlnet_hack(p)
      File "E:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\controlnet.py", line 1232, in controlnet_hack
        self.controlnet_main_entry(p)
      File "E:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\controlnet.py", line 982, in controlnet_main_entry
        controls, hr_controls, additional_maps = get_control(
      File "E:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\controlnet.py", line 233, in get_control
        input_image, resize_mode = Script.choose_input_image(p, unit, idx)
      File "E:\stable-diffusion-webui\extensions\sd-webui-controlnet\scripts\controlnet.py", line 793, in choose_input_image
        raise ValueError("controlnet is enabled but no input image is given")
    ValueError: controlnet is enabled but no input image is given
```

**Reply by huchenlei** [reply]
I don't think this is caused by advanced weighting. According to my local testing, advanced weighting can be used with multi-inputs.

This issue can be caused by improper sync of gr.State with UI state. This can happen if you do not reload the page after restart A1111 server.

**Comment by OoAnikioO** [comment]
what weight choice do you recommend for intant-id?

**Reply by TheSlayonetta** [reply]
I don't recommend weight choice lol it doesn't work with 1.5 and I can't run sdxl so idk

**Reply by huchenlei** [reply]
According to my test, for instant id, majority of work is done by the controlnet unit, so the advanced weighting on the ipadapter unit does not change the result much.

**Reply by huchenlei** [reply]
I might consider just remove advanced weighting for instant id later as it is not so useful.

**Reply by fugueblus** [reply]
I hope you can remove it, I think it has a big impact on the weight control of instantid，THX

**Comment by lleksa** [comment]
I'm using automatic1111 (with SD1.5 checkpoint model) with ControlNet v1.1.445 and there's no ip-adapter_clip_sd15 preprocessor. Is this normal? Thank you!

**Reply by huchenlei** [reply]
ip-adapter_clip_sd15 is alias of ip-adapter_clip_h. 

BTW, Now you can just use ip-adapter-auto to automatically select correct preprocessor for you.

**Comment by mykeehu** [comment]
I have noticed that if any of the prompts are 75 tokens longer, they lose the style of delivery. I tested with two models (babes and comicbabes), both of which lost style as soon as I went over 75 tokens.

55 token in negative prompt:
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/6390413/bb7ac9f1-033f-4dd4-b6c6-d266b1bfbef1)

74 token in negative prompt:
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/6390413/f614cafd-4070-4e02-92e0-256c056bc2d8)

76 token in negative prompt:
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/6390413/85d5dea4-67da-446d-9d1f-7f223b0dd838)

And 183 token in negative prompt:
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/6390413/3f8995dd-c546-432b-8109-216a3b1da6b4)

**Reply by huchenlei** [reply]
> IPA's decoupled Cross Attention training will be affected by the length of CLIP part
if their training is under standard CLIP only (77token
than the concat trick for CLIP will affect the attention significantly

Answer from A1111 dev.

**Comment by a3nima** [comment]
Any suggestions for settings with ipadapter face id v2 (sdxl) ? (normally I've been using weight 1), but with some source images the results are not good and i guess it could be better with other settings (lora 0.6 like suggested from dev).

**Reply by huangjianyi0701** [reply]
sdxl version of the weight can not be too high, lora weight can not be too high, probably all below 0.7, cfg adjusted to 3.5 or so, pro-tested to be useful, but by the way, I personally think that xl effect is not as good as 1.5

**Comment by TigerHH6866** [comment]
sometimes weight get wrong.
when i set to 0.1~0.9,all become 0
but 1 and above is ok
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/132218926/5347fab4-1914-42aa-834f-ed7588c658d5)

2024-05-15 10:50:57,852 - ControlNet - INFO - unit_separate = False, style_align = False█████████████████████████████████████████████████████████████████████████████████| 10/10 [00:02<00:00,  4.10it/s]
2024-05-15 10:50:57,852 - ControlNet - INFO - Loading model from cache: ip-adapter-faceid-plus_sd15 [d86a490f]
2024-05-15 10:50:57,887 - ControlNet - INFO - Using preprocessor: ip-adapter_face_id_plus
2024-05-15 10:50:57,887 - ControlNet - INFO - preprocessor resolution = 832
2024-05-15 10:50:57,890 - ControlNet - INFO - Loading model from cache: controlnet++_hed_softedge_sd15_fp16 [0d417fb5]
2024-05-15 10:50:57,921 - ControlNet - INFO - Using preprocessor: softedge_pidinet
2024-05-15 10:50:57,921 - ControlNet - INFO - preprocessor resolution = 832
2024-05-15 10:50:57,945 - ControlNet - INFO - IP-Adapter using advanced weighting [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
2024-05-15 10:50:57,997 - ControlNet - INFO - ControlNet Hooked - Time = 0.15296292304992676

**Reply by huchenlei** [reply]
I am not sure how did you get to that state but with following steps, it seems fine:
- select ipadapter control type
- select weight type ease in; observe that the weight plot shows
- select weight type normal: observe that the weight plot changes
- adjust control weight slider to a value < 0.9

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/d712ff2a-6291-461c-9d02-5b1d0857f075)

**Reply by yinian-lw** [reply]
same

**Comment by huanght18** [comment]
I'm using `ip-adapter_sd15_plus [32cd8f7f]` with preprocessor `ip-adapter_clip_h` in sd-webui, and I wonder what flow should I make in comfyui to achieve the same effect? Thanks a lot!

**Comment by INTstinkt** [comment]
Hey there, any chance to add "style transfer precise" and "composition precise" in the future?

**Comment by HorizonPursuer** [comment]
Can I use it in the API functionality? How to set the corresponding parameters?
